src/NHibernate/Cache/QueryCacheResultBuilder.cs

```diff
@@ -0,0 +1,120 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using NHibernate.Collection;
+using NHibernate.Engine;
+using NHibernate.Persister.Collection;
+using NHibernate.Type;
+
+namespace NHibernate.Cache
+{
+	/// <summary>
+	/// A builder that builds a list from a query that can be passed to <see cref="IBatchableQueryCache"/>.
+	/// </summary>
+	public sealed class QueryCacheResultBuilder
+	{
+		private readonly IType[] _resultTypes;
+		private readonly IType[] _cacheTypes;
+		private readonly List<int> _entityFetchIndexes = new List<int>();
+		private readonly List<int> _collectionFetchIndexes = new List<int>();
+		private readonly bool _hasFetches;
+
+		internal QueryCacheResultBuilder(Loader.Loader loader)
+		{
+			_resultTypes = loader.ResultTypes;
+			_cacheTypes = loader.CacheTypes;
+
+			if (loader.EntityFetches == null)
```

Comment:
This shortcut needs to be removed to support collection fetches without owner.

Reply:
Done.